File: python-wrapper/signavis/assets/widget.js
/**
 * anywidget ESM module for SignaVis.
 *
 * Communicates with the Python AudioWorkbenchWidget via the anywidget
 * model.get / model.set / model.on / model.send protocol.
 *
 * Trait sync (model ↔ Python):
 *   annotations, spectrogram_labels, current_time, duration, playing
 *
 * Custom messages (model.send / model.on("msg:custom")):
 *   Python → JS:  { type: "call", method, args }
 *   JS → Python:  { type: "event", event, detail }
 *
 * The IIFE bundle is prepended to this file by Python at import time,
 * so BirdNETPlayerModule is available as a module-scoped variable — no
 * eval() or script injection needed (works under strict CSP / Colab).
 */

function _ensureWaveSurfer() {
  return new Promise((resolve, reject) => {
    if (window.WaveSurfer) {
      resolve();
      return;
    }
    const script = document.createElement("script");
    script.src = "https://unpkg.com/wavesurfer@7";
    script.onload = resolve;
    script.onerror = () =>
      reject(new Error("Failed to load WaveSurfer.js from CDN"));
    document.head.appendChild(script);
  });
}

/**
 * Decode base64 audio string to Uint8Array.
 */
function _decodeBase64(b64) {
  const binary = atob(b64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
  return bytes;
}

/**
 * Serialize annotation regions to plain objects for Python sync.
 */
function _serializeAnnotations(player) {
  // player.annotations is the AnnotationLayer — it exposes .regions
  const layer = player.annotations;
  if (!layer || !layer.regions) return [];
  return layer.regions.map((r) => ({
    id: r.id,
    start: r.start,
    end: r.end,
    species: r.content?.textContent || r.species || "",
    color: r.color || "",
  }));
}

/**
 * Serialize spectrogram labels to plain objects for Python sync.
 */
function _serializeSpectrogramLabels(player) {
  const layer = player.spectrogramLabels;
  if (!layer || !layer.labels) return [];
  return layer.labels.map((l) => ({
    id: l.id,
    start: l.start,
    end: l.end,
    freqMin: l.freqMin,
    freqMax: l.freqMax,
    label: l.label || l.species || "",
    color: l.color || "",
  }));
}

/**
 * Push current player state to model traits.
 */
function _syncToModel(player, model) {
  model.set("annotations", _serializeAnnotations(player));
  model.set("spectrogram_labels", _serializeSpectrogramLabels(player));
  model.set("current_time", player.currentTime || 0);
  model.set("duration", player.duration || 0);
  model.save_changes();
}

// ── anywidget lifecycle ──────────────────────────────────────────────

export async function initialize({ model }) {
  // Load WaveSurfer from CDN (once)
  await _ensureWaveSurfer();
}

export async function render({ model, el }) {
  // ── DOM setup ──
  const container = document.createElement("div");
  container.style.width = "100%";
  container.style.height = "100%";
  el.appendChild(container);

  // Inject CSS
  const style = document.createElement("style");
  style.textContent = model.get("_player_css");
  el.appendChild(style);

  // ── Player options ──
  const playerOpts = JSON.parse(model.get("_player_options") || "{}");

  // ── Create player (BirdNETPlayerModule is module-scoped, prepended by Python) ──
  const BirdNETPlayer = BirdNETPlayerModule?.BirdNETPlayer;
  if (!BirdNETPlayer) {
    el.innerHTML =
      '<pre style="padding:12px;color:#b91c1c">BirdNETPlayer failed to load</pre>';
    return;
  }

  const player = new BirdNETPlayer(container, playerOpts);
  await player.ready;

  // ── Load audio ──
  const audioB64 = model.get("_audio_b64");
  if (audioB64) {
    const audioBytes = _decodeBase64(audioB64);
    const file = new File([audioBytes], "audio.wav", { type: "audio/wav" });
    await player.loadFile(file);
  }

  // ── Inject spectrogram (if provided) ──
  const spectB64 = model.get("_spectrogram_b64");
  const spectMeta = model.get("_spectrogram_meta");
  if (spectB64 && spectMeta) {
    const meta = JSON.parse(spectMeta);
    if (meta.type === "data") {
      const specBin = atob(spectB64);
      const specBytes = new Uint8Array(specBin.length);
      for (let i = 0; i < specBin.length; i++)
        specBytes[i] = specBin.charCodeAt(i);
      await player.setSpectrogramData(
        new Float32Array(specBytes.buffer),
        meta.n_frames,
        meta.n_mels,
        { mode: meta.mode || "perch", sampleRate: meta.sample_rate || 16000 }
      );
    } else if (meta.type === "image") {
      const imgSrc = meta.is_url
        ? spectB64
        : "data:image/png;base64," + spectB64;
      await player.setSpectrogramImage(imgSrc, {
        sampleRate: meta.sample_rate || undefined,
      });
    }
  }

  // ── Initial sync ──
  model.set("duration", player.duration || 0);
  model.set("playing", false);
  _syncToModel(player, model);

  // ── JS → Python: forward player events ──
  const SYNC_EVENTS = [
    "annotationcreate",
    "annotationupdate",
    "spectrogramlabelcreate",
    "spectrogramlabelupdate",
  ];

  for (const evt of SYNC_EVENTS) {
    player.on(evt, (detail) => {
      _syncToModel(player, model);
      model.send({ type: "event", event: evt, detail: detail || {} });
    });
  }

  player.on("timeupdate", (detail) => {
    model.set("current_time", detail?.currentTime ?? player.currentTime ?? 0);
    model.set("duration", detail?.duration ?? player.duration ?? 0);
    model.save_changes();
  });

  player.on("transportstatechange", (detail) => {
    const state = detail?.state || "";
    model.set("playing", state === "playing");
    model.save_changes();
    model.send({
      type: "event",
      event: "transportstatechange",
      detail: { state },
    });
  });

  // ── Python → JS: handle custom messages ──
  model.on("msg:custom", (msg) => {
    if (msg.type === "call" && typeof player[msg.method] === "function") {
      const args = msg.args || [];
      const result = player[msg.method](...args);
      // If it returns a promise, sync after resolution
      if (result && typeof result.then === "function") {
        result.then(() => _syncToModel(player, model));
      } else {
        _syncToModel(player, model);
      }
    } else if (msg.type === "sync") {
      _syncToModel(player, model);
    }
  });

  // ── Python → JS: react to trait changes ──
  model.on("change:_annotations_in", () => {
    const annots = model.get("_annotations_in");
    if (annots && Array.isArray(annots)) {
      player.setAnnotations(annots);
      _syncToModel(player, model);
    }
  });

  model.on("change:_spectrogram_labels_in", () => {
    const labels = model.get("_spectrogram_labels_in");
    if (labels && Array.isArray(labels)) {
      player.setSpectrogramLabels(labels);
      _syncToModel(player, model);
    }
  });

  // ── Cleanup ──
  return () => {
    player.destroy();
  };
}

File: python-wrapper/signavis/widget.py
"""anywidget-based SignaVis player with bidirectional communication.

Provides ``AudioWorkbenchWidget`` — a Jupyter widget that renders the
BirdNET DAW player and synchronizes annotations, labels, playback state,
and current time back to Python in real-time.
"""

from __future__ import annotations

import base64
import json
import pathlib
from typing import Any, Callable, Optional

import anywidget
import traitlets

from .renderer import (
    _PLAYER_CSS,
    _PLAYER_JS,
    _coerce_image_to_png_bytes,
    _encode_spectrogram_data,
)

_WIDGET_JS = (
    pathlib.Path(__file__).parent / "assets" / "widget.js"
).read_text(encoding="utf-8")

# Compose ESM: IIFE bundle (creates module-scoped BirdNETPlayerModule) +
# widget logic.  No eval() or script injection — works under strict CSP.
_WIDGET_ESM = _PLAYER_JS + "\n\n" + _WIDGET_JS


class AudioWorkbenchWidget(anywidget.AnyWidget):
    """Interactive SignaVis player widget for Jupyter notebooks.

    Provides bidirectional communication between Python and the JS player:

    * **Python → JS**: Call player methods via :meth:`call`, push annotations
      via :meth:`set_annotations`, push spectrogram labels via
      :meth:`set_spectrogram_labels`.
    * **JS → Python**: Read synced traits ``annotations``,
      ``spectrogram_labels``, ``current_time``, ``duration``, ``playing``.
      Register callbacks via :meth:`on_event`.

    Parameters
    ----------
    audio : bytes, str, or pathlib.Path
        Raw audio file content (WAV, MP3, OGG, etc.) or path to a file.
    spectrogram_data : numpy.ndarray, optional
        2D float32 array (nFrames × nMels) — player applies colorization.
    spectrogram_image : various, optional
        Pre-rendered image (bytes, str URL, PIL, matplotlib, numpy).
    spectrogram_mode : str
        ``'perch'`` or ``'classic'``. Default: ``'perch'``.
    sample_rate : int, optional
        Override sample rate for frequency axis.
    **options
        Forwarded to the ``BirdNETPlayer`` JS constructor (e.g.
        ``viewMode='spectrogram'``, ``showFileOpen=False``).

    Examples
    --------
    >>> from signavis import AudioWorkbenchWidget
    >>> w = AudioWorkbenchWidget(audio_bytes, viewMode="both")
    >>> w  # display in notebook cell
    >>> # After user draws annotations:
    >>> w.annotations
    [{'id': '...', 'start': 1.2, 'end': 2.5, 'species': 'Bird Call', ...}]
    >>> w.spectrogram_labels
    [{'id': '...', 'start': 1.0, 'end': 3.0, 'freqMin': 2000, ...}]
    """

    # ── Internal traits (not meant for direct user access) ──
    _esm = traitlets.Unicode().tag(sync=True)
    _css = traitlets.Unicode("").tag(sync=True)

    _player_js = traitlets.Unicode("").tag(sync=True)
    _player_css = traitlets.Unicode("").tag(sync=True)
    _player_options = traitlets.Unicode("{}").tag(sync=True)
    _audio_b64 = traitlets.Unicode("").tag(sync=True)
    _spectrogram_b64 = traitlets.Unicode("").tag(sync=True)
    _spectrogram_meta = traitlets.Unicode("").tag(sync=True)

    # Python → JS push traits
    _annotations_in = traitlets.List(traitlets.Dict(), default_value=[]).tag(
        sync=True
    )
    _spectrogram_labels_in = traitlets.List(
        traitlets.Dict(), default_value=[]
    ).tag(sync=True)

    # ── JS → Python synced traits (read from Python) ──
    annotations = traitlets.List(traitlets.Dict(), default_value=[]).tag(
        sync=True
    )
    spectrogram_labels = traitlets.List(
        traitlets.Dict(), default_value=[]
    ).tag(sync=True)
    current_time = traitlets.Float(0.0).tag(sync=True)
    duration = traitlets.Float(0.0).tag(sync=True)
    playing = traitlets.Bool(False).tag(sync=True)

    def __init__(
        self,
        audio,
        *,
        spectrogram_data=None,
        spectrogram_image=None,
        spectrogram_mode: str = "perch",
        sample_rate: Optional[int] = None,
        **options,
    ):
        if spectrogram_data is not None and spectrogram_image is not None:
            raise ValueError(
                "Provide spectrogram_data OR spectrogram_image, not both."
            )

        # Accept file paths
        if isinstance(audio, (str, pathlib.Path)):
            audio = pathlib.Path(audio).read_bytes()

        # Encode audio
        audio_b64 = base64.b64encode(audio).decode("ascii")

        # Encode spectrogram
        spect_b64 = ""
        spect_meta = ""
        if spectrogram_data is not None:
            b64_data, n_frames, n_mels = _encode_spectrogram_data(
                spectrogram_data
            )
            spect_b64 = b64_data
            spect_meta = json.dumps(
                {
                    "type": "data",
                    "n_frames": n_frames,
                    "n_mels": n_mels,
                    "mode": spectrogram_mode,
                    "sample_rate": sample_rate or 16000,
                }
            )
        elif spectrogram_image is not None:
            if isinstance(spectrogram_image, str):
                spect_b64 = spectrogram_image
                spect_meta = json.dumps(
                    {
                        "type": "image",
                        "is_url": True,
                        "sample_rate": sample_rate,
                    }
                )
            else:
                png_bytes = _coerce_image_to_png_bytes(spectrogram_image)
                spect_b64 = base64.b64encode(png_bytes).decode("ascii")
                spect_meta = json.dumps(
                    {
                        "type": "image",
                        "is_url": False,
                        "sample_rate": sample_rate,
                    }
                )

        super().__init__(
            _esm=_WIDGET_ESM,
            _player_css=_PLAYER_CSS,
            _player_options=json.dumps(options or {}),
            _audio_b64=audio_b64,
            _spectrogram_b64=spect_b64,
            _spectrogram_meta=spect_meta,
        )

        self._event_callbacks: dict[str, list[Callable]] = {}
        self.on_msg(self._handle_custom_msg)

    # ── Python → JS methods ──

    def call(self, method: str, *args: Any) -> None:
        """Call a method on the JS BirdNETPlayer instance.

        Parameters
        ----------
        method : str
            Name of the player method (e.g. ``'play'``, ``'pause'``,
            ``'stop'``, ``'addAnnotation'``, ``'clearAnnotations'``).
        *args
            Arguments forwarded to the JS method.

        Examples
        --------
        >>> w.call('play')
        >>> w.call('pause')
        >>> w.call('addAnnotation', {'start': 1.0, 'end': 2.0, 'species': 'Bird'})
        """
        self.send({"type": "call", "method": method, "args": list(args)})

    def play(self) -> None:
        """Start playback."""
        self.call("play")

    def pause(self) -> None:
        """Pause playback."""
        self.call("pause")

    def stop(self) -> None:
        """Stop playback."""
        self.call("stop")

    def set_annotations(self, annotations: list[dict]) -> None:
        """Replace all waveform annotations from Python.

        Parameters
        ----------
        annotations : list of dict
            Each dict should have ``start``, ``end``, and optionally
            ``species``, ``color``, ``confidence``.
        """
        self._annotations_in = list(annotations)

    def set_spectrogram_labels(self, labels: list[dict]) -> None:
        """Replace all spectrogram labels from Python.

        Parameters
        ----------
        labels : list of dict
            Each dict should have ``start``, ``end``, ``freqMin``, ``freqMax``,
            and optionally ``label``, ``color``.
        """
        self._spectrogram_labels_in = list(labels)

    def export_annotations_raven(self) -> None:
        """Trigger Raven-format annotation export on the JS side."""
        self.call("exportAnnotationsRaven")

    def sync(self) -> None:
        """Request a full state sync from JS → Python."""
        self.send({"type": "sync"})

    # ── Event system ──

    def on_event(self, event: str, callback: Callable) -> None:
        """Register a callback for a JS player event.

        Parameters
        ----------
        event : str
            Event name (e.g. ``'annotationcreate'``,
            ``'spectrogramlabelupdate'``, ``'transportstatechange'``).
        callback : callable
            Called with ``(event_name: str, detail: dict)`` when fired.

        Examples
        --------
        >>> def on_annotation(event, detail):
        ...     print(f"New annotation: {detail}")
        >>> w.on_event('annotationcreate', on_annotation)
        """
        self._event_callbacks.setdefault(event, []).append(callback)

    def _handle_custom_msg(self, msg, buffers):
        """Dispatch incoming custom messages from JS."""
        if not isinstance(msg, dict):
            return
        if msg.get("type") == "event":
            event = msg.get("event", "")
            detail = msg.get("detail", {})
            for cb in self._event_callbacks.get(event, []):
                cb(event, detail)
